File: TyreVibes/Core/Service/PlateAPIService.swift
import Foundation
import SwiftUI
import Supabase

struct PlateAPIRequest: Codable {
    let plate: String
    let make: String?
    let model: String?
    let color: String?
    let fuelType: String?
    let powerKW: String?
    let powerCV: String?
    let modelDetails: String?
    let displacementCC: String?
    let registrationDate: String?
    let year: Int?
    let month: Int?
    let vin: String?

    let userId: String
    let imagesBase64: [String]?
    let imagesMime: [String]?
    let imagesAngle: [Int]?
    let imagesColor: [String]

    // RCA/Assicurazione
    let insuranceCompany: String?
    let insurancePolicyNumber: String?
    let insuranceExpiry: Date?
    let insurancePresent: Bool?

    let emissionClass: String?
    let tyres: [[String: String]]?

    // --- campi da /dettagli ---
    let view: String?
    let saleStart: String?
    let saleEnd: String?
    let gearbox: String?
    let maxSpeed: String?
    let bodyType: String?
    let doors: String?
    let seats: String?
    let consumption: String?
    let traction: String?

    let revisioni: [Revisione]?

    // Bollo
    let bolloAmount: Double?
    let bolloSuper: Double?
    let bolloTotal: Double?
    let bolloEmissionClass: String?
}

struct PlateAPIResponse: Codable {
    let plate_id: Int
    let plate: String
    let make: String?
    let model: String?
    let modelDetail : String?
    let fuel_type: String?
    let power_kw: String?
    let cilindrata: String?
    let color: String?
    let vin: String?
    let created_at: String
    let user_id: String?
    let registration_date: String?
    let image_base64: String?
    
    let rcaCompany: String?
    let rcaPolicyNumber: String?
    let rcaExpiry: Date?
    let rcaInsurancePresent: Bool?
    let classeAmbientale: String?
}

enum PlateAPIError: Error {
    case invalidURL
    case requestFailed(Error)
    case invalidResponse
    case serverError(Int, String)
    case plateNotFound
    case alreadyInGarage
}

private func imageHasAlpha(_ image: UIImage) -> Bool {
    guard let alpha = image.cgImage?.alphaInfo else { return false }
    switch alpha {
    case .first, .last, .premultipliedFirst, .premultipliedLast:
        return true
    default:
        return false
    }
}

class PlateAPIService {

    static let apiConfig: NSDictionary = {
        if let path = Bundle.main.path(forResource: "Api", ofType: "plist"),
           let dict = NSDictionary(contentsOfFile: path) {
            return dict
        }
        return [:]
    }()

    private let savePlateURL = URL(string: (PlateAPIService.apiConfig["SavePlateURL"] as? String) ?? "")
    private let checkPlateBaseURL = (PlateAPIService.apiConfig["CheckPlateBaseURL"] as? String) ?? ""
    private let manualPlateURL = URL(string: (PlateAPIService.apiConfig["ManualPlateURL"] as? String) ?? "")

    // MARK: - Get Supabase JWT Token
    private func getAuthToken() async -> String? {
        do {
            let session = try await SupabaseManager.client.auth.session
            return session.accessToken
        } catch {
            print("⚠️ [PlateAPIService] Failed to get auth token: \(error.localizedDescription)")
            return nil
        }
    }

    static var currentUserToken: String? {
        get async {
            do {
                let session = try await SupabaseManager.client.auth.session
                return session.accessToken
            } catch {
                return nil
            }
        }
    }
    
    private func convertRegistrationDate(_ dateStr: String?) -> String {
        guard let dateStr = dateStr else { return "-" }
        let comps = dateStr.split(separator: "/")
        if comps.count == 2, let month = comps.first, let year = comps.last {
            return "\(year)-\(month)-01"
        }
        return "-"
    }

    func extractYear(_ dateStr: String?) -> Int? {
        guard let dateStr = dateStr else { return nil }
        let formatterInput = DateFormatter()
        formatterInput.dateFormat = "MM/yyyy"
        formatterInput.locale = Locale(identifier: "it_IT_POSIX")
        if let date = formatterInput.date(from: dateStr) {
            let calendar = Calendar.current
            return calendar.component(.year, from: date)
        }
        return nil
    }

    private func extractMonth(_ dateStr: String?) -> Int? {
        guard let dateStr = dateStr else { return nil }
        let formatterInput = DateFormatter()
        formatterInput.dateFormat = "MM/yyyy"
        formatterInput.locale = Locale(identifier: "it_IT_POSIX")
        if let date = formatterInput.date(from: dateStr) {
            let calendar = Calendar.current
            return calendar.component(.month, from: date)
        }
        return nil
    }
    
    
    
    
    func checkPlate(plateNumber: String) async throws -> Int? {
        guard var components = URLComponents(string: checkPlateBaseURL) else {
            throw PlateAPIError.invalidURL
        }
        components.queryItems = [URLQueryItem(name: "plate", value: plateNumber)]

        guard let url = components.url else {
            throw PlateAPIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        // Add JWT token
        if let token = await getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw PlateAPIError.invalidResponse
        }
        
        if httpResponse.statusCode == 404 {
            return nil // Plate not found is not an error, it's a valid outcome.
        }
        
        if (200...299).contains(httpResponse.statusCode) {
            do {
                let apiResponse = try JSONDecoder().decode(PlateAPIResponse.self, from: data)
                return apiResponse.plate_id
            } catch {
                throw PlateAPIError.requestFailed(error) // JSON decoding error
            }
        } else {
            let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown server error"
            throw PlateAPIError.serverError(httpResponse.statusCode, errorMessage)
        }
    }
    
    func savePlate(plateData: PlateData, color: String, userId: String, images: [UIImage?], imagesColor: [String]) async throws {
        guard let url = savePlateURL else {
            throw PlateAPIError.invalidURL
        }

        var imagesBase64: [String] = []
        var imagesMime: [String] = []
        for image in images {
            if let uiImage = image {
                if imageHasAlpha(uiImage), let data = uiImage.pngData() {
                    imagesBase64.append(data.base64EncodedString())
                    imagesMime.append("image/png")
                } else if let data = uiImage.jpegData(compressionQuality: 0.9) {
                    imagesBase64.append(data.base64EncodedString())
                    imagesMime.append("image/jpeg")
                }
            }
        }

        let angles: [Int]
        if imagesBase64.count == 2 {
            angles = Array([23, 12])
        } else {
            angles = Array([23,23,12,12])
        }

        let requestBody = PlateAPIRequest(
            plate: plateData.plate.uppercased(),
            make: plateData.make ?? "-",
            model: plateData.model ?? "-",
            color: color.lowercased(),
            fuelType: plateData.fuelType ?? "-",
            powerKW: plateData.powerKW ?? "-",
            powerCV: plateData.powerCV ?? "-",
            modelDetails: plateData.modelDetails ?? "-",
            displacementCC: plateData.displacementCC ?? "-",
            registrationDate: convertRegistrationDate(plateData.registrationDate),
            year: extractYear(plateData.registrationDate),
            month: extractMonth(plateData.registrationDate),
            vin: plateData.vin ?? "-",
            userId: userId,
            imagesBase64: imagesBase64,
            imagesMime: imagesMime,
            imagesAngle: angles,
            imagesColor: imagesColor,
            insuranceCompany: plateData.insuranceCompany,
            insurancePolicyNumber: plateData.insurancePolicyNumber,
            insuranceExpiry: plateData.insuranceExpiry,
            insurancePresent: plateData.insurancePresent,
            emissionClass: plateData.emissionClass,
            tyres: plateData.tyres,
            view: plateData.view,
            saleStart : plateData.saleStart,
            saleEnd: plateData.saleEnd,
            gearbox: plateData.gearbox,
            maxSpeed: plateData.maxSpeed,
            bodyType: plateData.bodyType,
            doors: plateData.doors,
            seats: plateData.seats,
            consumption: plateData.consumption,
            traction: plateData.traction,
            revisioni: plateData.revisioni ?? [],
            bolloAmount: plateData.bollo?.baseBollo,
            bolloSuper: plateData.bollo?.superBollo,
            bolloTotal: plateData.bollo?.total,
            bolloEmissionClass: plateData.bollo?.emissionClassDescription
        )
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Add JWT token
        if let token = await getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        do {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            request.httpBody = try encoder.encode(requestBody)
        } catch {
            throw PlateAPIError.requestFailed(error)
        }
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw PlateAPIError.invalidResponse
        }
        
        if (200...299).contains(httpResponse.statusCode) {
            // Success
            return
        } else {
            // Server error
            let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown server error"
            throw PlateAPIError.serverError(httpResponse.statusCode, errorMessage)
        }
    }
    
    func savePlateManually(
        plate: String,
        make: String = "",
        model: String = "",
        registrationDate: String = "",
        fuelType: String = "",
        powerKW: String = "",
        powerCV: String = "",
        displacementCC: String = "",
        emissionClass: String = "",
        gearbox: String = "",
        maxSpeed: String = "",
        bodyType: String = "",
        doors: String = "",
        seats: String = "",
        consumption: String = "",
        traction: String = "",
        saleStart: String = "",
        saleEnd: String = "",
        color: String = "",
        vin: String = "",
        modelDetails: String = "",
        insuranceCompany: String? = nil,
        insurancePolicyNumber: String? = nil,
        insuranceExpiry: Date? = nil,
        insurancePresent: Bool? = nil,
        userId: String
    ) async throws -> (vehicleId: Int, plateId: Int) {
        guard let url = manualPlateURL else {
            throw PlateAPIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Add JWT token
        if let token = await getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let requestBody: [String: Any] = [
            "plate": plate.uppercased(),
            "make": make.isEmpty ? "" : make,
            "model": model.isEmpty ? "" : model,
            "modelDetails": modelDetails.isEmpty ? "" : modelDetails,
            "registrationDate": registrationDate.isEmpty ? "" : registrationDate,
            "fuelType": fuelType.isEmpty ? "" : fuelType,
            "powerKW": powerKW.isEmpty ? "" : powerKW,
            "powerCV": powerCV.isEmpty ? "" : powerCV,
            "displacementCC": displacementCC.isEmpty ? "" : displacementCC,
            "emissionClass": emissionClass.isEmpty ? "" : emissionClass,
            "gearbox": gearbox.isEmpty ? "" : gearbox,
            "maxSpeed": maxSpeed.isEmpty ? "" : maxSpeed,
            "bodyType": bodyType.isEmpty ? "" : bodyType,
            "doors": doors.isEmpty ? "" : doors,
            "seats": seats.isEmpty ? "" : seats,
            "consumption": consumption.isEmpty ? "" : consumption,
            "traction": traction.isEmpty ? "" : traction,
            "saleStart": saleStart.isEmpty ? "" : saleStart,
            "saleEnd": saleEnd.isEmpty ? "" : saleEnd,
            "color": color.isEmpty ? "" : color,
            "vin": vin.isEmpty ? "" : vin,
            "insuranceCompany": insuranceCompany ?? NSNull(),
            "insurancePolicyNumber": insurancePolicyNumber ?? NSNull(),
            "insuranceExpiry": insuranceExpiry ?? NSNull(),
            "insurancePresent": insurancePresent ?? NSNull(),
            "userId": userId
        ]

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody, options: [])
        } catch {
            throw PlateAPIError.requestFailed(error)
        }

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw PlateAPIError.invalidResponse
        }

        if (200...299).contains(httpResponse.statusCode) {
            struct ManualPlateResponse: Codable {
                let vehicle_id: Int
                let plate_id: Int
                let message: String
            }

            do {
                let decoded = try JSONDecoder().decode(ManualPlateResponse.self, from: data)
                return (decoded.vehicle_id, decoded.plate_id)
            } catch {
                throw PlateAPIError.requestFailed(error)
            }
        } else {
            let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown server error"
            throw PlateAPIError.serverError(httpResponse.statusCode, errorMessage)
        }
    }

    func associateVehicle2User(vehicleId: Int, userId: String, color: String = "", images: [UIImage]? = nil) async throws -> (message: String, imageBase64: String?) {
        guard let baseURLString = PlateAPIService.apiConfig["BASE_URL"] as? String else {
            throw PlateAPIError.invalidURL
        }
        // Remove trailing slash if present
        let urlString = "\(baseURLString)/v1/vehicles/\(vehicleId)/user/\(userId)"

        guard let url = URL(string: urlString) else {
            throw PlateAPIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Add JWT token
        if let token = await getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        // Prepara il body della richiesta
        struct RequestBody: Encodable {
            let color: String
            let imagesBase64: [String]?
            let imagesMime: [String]?
            let imagesAngle: [Int]?
        }

        var imagesBase64: [String]? = nil
        var imagesMime: [String]? = nil
        var imagesAngle: [Int]? = nil

        // Se ci sono immagini, convertile in base64
        if let images = images {
            var base64Images: [String] = []
            var mimeTypes: [String] = []

            for image in images {
                if imageHasAlpha(image), let data = image.pngData() {
                    base64Images.append(data.base64EncodedString())
                    mimeTypes.append("image/png")
                } else if let data = image.jpegData(compressionQuality: 0.9) {
                    base64Images.append(data.base64EncodedString())
                    mimeTypes.append("image/jpeg")
                }
            }

            imagesBase64 = base64Images
            imagesMime = mimeTypes
            imagesAngle = [23, 23, 12, 12] // angoli fissi per le 4 immagini
        }

        let body = RequestBody(
            color: color,
            imagesBase64: imagesBase64,
            imagesMime: imagesMime,
            imagesAngle: imagesAngle
        )

        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw PlateAPIError.invalidResponse
        }

        if (200...299).contains(httpResponse.statusCode) {
            struct AssociationResponse: Codable {
                let message: String
                let image_base64: String?
            }
            let decoded = try JSONDecoder().decode(AssociationResponse.self, from: data)
            return (decoded.message, decoded.image_base64)
        } else {
            let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown server error"
            throw PlateAPIError.serverError(httpResponse.statusCode, errorMessage)
        }
    }
}

File: android/app/src/main/java/it/tyrevibes/app/core/service/PlateAPIService.kt
package it.tyrevibes.app.core.service

import android.content.Context
import android.graphics.Bitmap
import android.util.Base64
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import it.tyrevibes.app.core.model.Revisione
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.io.ByteArrayOutputStream
import java.text.SimpleDateFormat
import java.util.*

/**
 * Plate Data - Data for plate registration
 */
data class PlateData(
    val plate: String,
    val make: String? = null,
    val model: String? = null,
    val color: String? = null,
    val fuelType: String? = null,
    val powerKW: String? = null,
    val powerCV: String? = null,
    val modelDetails: String? = null,
    val displacementCC: String? = null,
    val registrationDate: String? = null,
    val vin: String? = null,
    val insuranceCompany: String? = null,
    val insurancePolicyNumber: String? = null,
    val insuranceExpiry: Date? = null,
    val insurancePresent: Boolean? = null,
    val emissionClass: String? = null,
    val tyres: List<Map<String, String>>? = null,
    val view: String? = null,
    val saleStart: String? = null,
    val saleEnd: String? = null,
    val gearbox: String? = null,
    val maxSpeed: String? = null,
    val bodyType: String? = null,
    val doors: String? = null,
    val seats: String? = null,
    val consumption: String? = null,
    val traction: String? = null,
    val revisioni: List<Revisione>? = null,
    val bollo: BolloInfo? = null
)

data class BolloInfo(
    val baseBollo: Double?,
    val superBollo: Double?,
    val total: Double?,
    val emissionClassDescription: String?
)

/**
 * Plate API Request model
 */
@Serializable
data class PlateAPIRequest(
    val plate: String,
    val make: String?,
    val model: String?,
    val color: String?,
    val fuelType: String?,
    val powerKW: String?,
    val powerCV: String?,
    val modelDetails: String?,
    val displacementCC: String?,
    val registrationDate: String?,
    val year: Int?,
    val month: Int?,
    val vin: String?,
    val userId: String,
    val imagesBase64: List<String>?,
    val imagesMime: List<String>?,
    val imagesAngle: List<Int>?,
    val imagesColor: List<String>,
    val insuranceCompany: String?,
    val insurancePolicyNumber: String?,
    val insuranceExpiry: String?,
    val insurancePresent: Boolean?,
    val emissionClass: String?,
    val tyres: List<Map<String, String>>?,
    val view: String?,
    val saleStart: String?,
    val saleEnd: String?,
    val gearbox: String?,
    val maxSpeed: String?,
    val bodyType: String?,
    val doors: String?,
    val seats: String?,
    val consumption: String?,
    val traction: String?,
    val revisioni: List<Revisione>?,
    val bolloAmount: Double?,
    val bolloSuper: Double?,
    val bolloTotal: Double?,
    val bolloEmissionClass: String?
)

/**
 * Plate API Response model
 */
@Serializable
data class PlateAPIResponse(
    @SerialName("plate_id") val plateId: Int,
    val plate: String,
    val make: String?,
    val model: String?,
    val modelDetail: String?,
    @SerialName("fuel_type") val fuelType: String?,
    @SerialName("power_kw") val powerKw: String?,
    val cilindrata: String?,
    val color: String?,
    val vin: String?,
    @SerialName("created_at") val createdAt: String,
    @SerialName("user_id") val userId: String?,
    @SerialName("registration_date") val registrationDate: String?,
    @SerialName("image_base64") val imageBase64: String?,
    val rcaCompany: String?,
    val rcaPolicyNumber: String?,
    val rcaExpiry: String?,
    val rcaInsurancePresent: Boolean?,
    val classeAmbientale: String?
)

/**
 * Manual Plate Response
 */
@Serializable
data class ManualPlateResponse(
    @SerialName("vehicle_id") val vehicleId: Int,
    @SerialName("plate_id") val plateId: Int,
    val message: String
)

/**
 * Association Response
 */
@Serializable
data class AssociationResponse(
    val message: String,
    @SerialName("image_base64") val imageBase64: String?
)

/**
 * Plate API Errors
 */
sealed class PlateAPIError : Exception() {
    object InvalidURL : PlateAPIError()
    data class RequestFailed(val error: Throwable) : PlateAPIError()
    object InvalidResponse : PlateAPIError()
    data class ServerError(val statusCode: Int, val message: String) : PlateAPIError()
    object PlateNotFound : PlateAPIError()
    object AlreadyInGarage : PlateAPIError()
}

/**
 * Plate API Service - Handles external plate API calls
 */
class PlateAPIService(
    private val context: Context,
    private val httpClient: HttpClient,
    private val supabaseManager: SupabaseManager
) {
    // TODO: Load from configuration file or BuildConfig
    private val savePlateURL = "https://your-api-url/save-plate" // Replace with actual URL
    private val checkPlateBaseURL = "https://your-api-url/check-plate" // Replace with actual URL
    private val manualPlateURL = "https://your-api-url/manual-plate" // Replace with actual URL
    private val baseURL = "https://your-api-url" // Replace with actual URL

    /**
     * Get auth token from Supabase
     */
    private suspend fun getAuthToken(): String? {
        return try {
            supabaseManager.getAccessToken()
        } catch (e: Exception) {
            println("⚠️ [PlateAPIService] Failed to get auth token: ${e.message}")
            null
        }
    }

    /**
     * Check if plate exists in database
     */
    suspend fun checkPlate(plateNumber: String): Int? {
        return try {
            val token = getAuthToken() ?: throw PlateAPIError.InvalidResponse

            val response: HttpResponse = httpClient.get(checkPlateBaseURL) {
                parameter("plate", plateNumber)
                header(HttpHeaders.Authorization, "Bearer $token")
            }

            when (response.status.value) {
                404 -> null // Plate not found
                in 200..299 -> {
                    val apiResponse = response.body<PlateAPIResponse>()
                    apiResponse.plateId
                }
                else -> {
                    val errorMessage = response.bodyAsText()
                    throw PlateAPIError.ServerError(response.status.value, errorMessage)
                }
            }
        } catch (e: Exception) {
            when (e) {
                is PlateAPIError -> throw e
                else -> throw PlateAPIError.RequestFailed(e)
            }
        }
    }

    /**
     * Convert registration date MM/yyyy to yyyy-MM-dd
     */
    private fun convertRegistrationDate(dateStr: String?): String {
        dateStr ?: return "-"
        val parts = dateStr.split("/")
        return if (parts.size == 2) {
            "${parts[1]}-${parts[0]}-01"
        } else {
            "-"
        }
    }

    /**
     * Extract year from MM/yyyy format
     */
    private fun extractYear(dateStr: String?): Int? {
        dateStr ?: return null
        return try {
            val formatter = SimpleDateFormat("MM/yyyy", Locale.ITALIAN)
            val date = formatter.parse(dateStr)
            val calendar = Calendar.getInstance()
            calendar.time = date!!
            calendar.get(Calendar.YEAR)
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Extract month from MM/yyyy format
     */
    private fun extractMonth(dateStr: String?): Int? {
        dateStr ?: return null
        return try {
            val formatter = SimpleDateFormat("MM/yyyy", Locale.ITALIAN)
            val date = formatter.parse(dateStr)
            val calendar = Calendar.getInstance()
            calendar.time = date!!
            calendar.get(Calendar.MONTH) + 1 // Calendar.MONTH is 0-based
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Convert Bitmap to Base64
     */
    private fun bitmapToBase64(bitmap: Bitmap): Pair<String, String> {
        val byteArrayOutputStream = ByteArrayOutputStream()

        // Check if bitmap has alpha channel
        val hasAlpha = bitmap.hasAlpha()

        return if (hasAlpha) {
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream)
            val byteArray = byteArrayOutputStream.toByteArray()
            Pair(Base64.encodeToString(byteArray, Base64.NO_WRAP), "image/png")
        } else {
            bitmap.compress(Bitmap.CompressFormat.JPEG, 90, byteArrayOutputStream)
            val byteArray = byteArrayOutputStream.toByteArray()
            Pair(Base64.encodeToString(byteArray, Base64.NO_WRAP), "image/jpeg")
        }
    }

    /**
     * Save plate with images
     */
    suspend fun savePlate(
        plateData: PlateData,
        color: String,
        userId: String,
        images: List<Bitmap?>,
        imagesColor: List<String>
    ) {
        val token = getAuthToken() ?: throw PlateAPIError.InvalidResponse

        val imagesBase64 = mutableListOf<String>()
        val imagesMime = mutableListOf<String>()

        images.filterNotNull().forEach { bitmap ->
            val (base64, mime) = bitmapToBase64(bitmap)
            imagesBase64.add(base64)
            imagesMime.add(mime)
        }

        val angles = if (imagesBase64.size == 2) {
            listOf(23, 12)
        } else {
            listOf(23, 23, 12, 12)
        }

        val requestBody = PlateAPIRequest(
            plate = plateData.plate.uppercase(),
            make = plateData.make ?: "-",
            model = plateData.model ?: "-",
            color = color.lowercase(),
            fuelType = plateData.fuelType ?: "-",
            powerKW = plateData.powerKW ?: "-",
            powerCV = plateData.powerCV ?: "-",
            modelDetails = plateData.modelDetails ?: "-",
            displacementCC = plateData.displacementCC ?: "-",
            registrationDate = convertRegistrationDate(plateData.registrationDate),
            year = extractYear(plateData.registrationDate),
            month = extractMonth(plateData.registrationDate),
            vin = plateData.vin ?: "-",
            userId = userId,
            imagesBase64 = imagesBase64,
            imagesMime = imagesMime,
            imagesAngle = angles,
            imagesColor = imagesColor,
            insuranceCompany = plateData.insuranceCompany,
            insurancePolicyNumber = plateData.insurancePolicyNumber,
            insuranceExpiry = plateData.insuranceExpiry?.toString(),
            insurancePresent = plateData.insurancePresent,
            emissionClass = plateData.emissionClass,
            tyres = plateData.tyres,
            view = plateData.view,
            saleStart = plateData.saleStart,
            saleEnd = plateData.saleEnd,
            gearbox = plateData.gearbox,
            maxSpeed = plateData.maxSpeed,
            bodyType = plateData.bodyType,
            doors = plateData.doors,
            seats = plateData.seats,
            consumption = plateData.consumption,
            traction = plateData.traction,
            revisioni = plateData.revisioni,
            bolloAmount = plateData.bollo?.baseBollo,
            bolloSuper = plateData.bollo?.superBollo,
            bolloTotal = plateData.bollo?.total,
            bolloEmissionClass = plateData.bollo?.emissionClassDescription
        )

        try {
            val response: HttpResponse = httpClient.post(savePlateURL) {
                header(HttpHeaders.Authorization, "Bearer $token")
                contentType(ContentType.Application.Json)
                setBody(requestBody)
            }

            if (response.status.value !in 200..299) {
                val errorMessage = response.bodyAsText()
                throw PlateAPIError.ServerError(response.status.value, errorMessage)
            }
        } catch (e: Exception) {
            when (e) {
                is PlateAPIError -> throw e
                else -> throw PlateAPIError.RequestFailed(e)
            }
        }
    }

    /**
     * Save plate manually (without images)
     */
    suspend fun savePlateManually(
        plate: String,
        make: String = "",
        model: String = "",
        modelDetails: String = "",
        registrationDate: String = "",
        fuelType: String = "",
        powerKW: String = "",
        powerCV: String = "",
        displacementCC: String = "",
        emissionClass: String = "",
        gearbox: String = "",
        maxSpeed: String = "",
        bodyType: String = "",
        doors: String = "",
        seats: String = "",
        consumption: String = "",
        traction: String = "",
        saleStart: String = "",
        saleEnd: String = "",
        color: String = "",
        vin: String = "",
        insuranceCompany: String? = null,
        insurancePolicyNumber: String? = null,
        insuranceExpiry: Date? = null,
        insurancePresent: Boolean? = null,
        userId: String
    ): Pair<Int, Int> {
        val token = getAuthToken() ?: throw PlateAPIError.InvalidResponse

        val requestBody = mapOf(
            "plate" to plate.uppercase(),
            "make" to make.ifEmpty { "" },
            "model" to model.ifEmpty { "" },
            "modelDetails" to modelDetails.ifEmpty { "" },
            "registrationDate" to registrationDate.ifEmpty { "" },
            "fuelType" to fuelType.ifEmpty { "" },
            "powerKW" to powerKW.ifEmpty { "" },
            "powerCV" to powerCV.ifEmpty { "" },
            "displacementCC" to displacementCC.ifEmpty { "" },
            "emissionClass" to emissionClass.ifEmpty { "" },
            "gearbox" to gearbox.ifEmpty { "" },
            "maxSpeed" to maxSpeed.ifEmpty { "" },
            "bodyType" to bodyType.ifEmpty { "" },
            "doors" to doors.ifEmpty { "" },
            "seats" to seats.ifEmpty { "" },
            "consumption" to consumption.ifEmpty { "" },
            "traction" to traction.ifEmpty { "" },
            "saleStart" to saleStart.ifEmpty { "" },
            "saleEnd" to saleEnd.ifEmpty { "" },
            "color" to color.ifEmpty { "" },
            "vin" to vin.ifEmpty { "" },
            "insuranceCompany" to (insuranceCompany ?: ""),
            "insurancePolicyNumber" to (insurancePolicyNumber ?: ""),
            "insuranceExpiry" to (insuranceExpiry?.toString() ?: ""),
            "insurancePresent" to (insurancePresent ?: false),
            "userId" to userId
        )

        return try {
            val response: HttpResponse = httpClient.post(manualPlateURL) {
                header(HttpHeaders.Authorization, "Bearer $token")
                contentType(ContentType.Application.Json)
                setBody(requestBody)
            }

            if (response.status.value in 200..299) {
                val decoded = response.body<ManualPlateResponse>()
                Pair(decoded.vehicleId, decoded.plateId)
            } else {
                val errorMessage = response.bodyAsText()
                throw PlateAPIError.ServerError(response.status.value, errorMessage)
            }
        } catch (e: Exception) {
            when (e) {
                is PlateAPIError -> throw e
                else -> throw PlateAPIError.RequestFailed(e)
            }
        }
    }

    /**
     * Associate vehicle to user
     */
    suspend fun associateVehicle2User(
        vehicleId: Int,
        userId: String,
        color: String = "",
        images: List<Bitmap>? = null
    ): Pair<String, String?> {
        val token = getAuthToken() ?: throw PlateAPIError.InvalidResponse
        val url = "$baseURL/v1/vehicles/$vehicleId/user/$userId"

        var imagesBase64: List<String>? = null
        var imagesMime: List<String>? = null
        var imagesAngle: List<Int>? = null

        images?.let { bitmaps ->
            val base64List = mutableListOf<String>()
            val mimeList = mutableListOf<String>()

            bitmaps.forEach { bitmap ->
                val (base64, mime) = bitmapToBase64(bitmap)
                base64List.add(base64)
                mimeList.add(mime)
            }

            imagesBase64 = base64List
            imagesMime = mimeList
            imagesAngle = listOf(23, 23, 12, 12)
        }

        val requestBody = mapOf(
            "color" to color,
            "imagesBase64" to imagesBase64,
            "imagesMime" to imagesMime,
            "imagesAngle" to imagesAngle
        )

        return try {
            val response: HttpResponse = httpClient.post(url) {
                header(HttpHeaders.Authorization, "Bearer $token")
                contentType(ContentType.Application.Json)
                setBody(requestBody)
            }

            if (response.status.value in 200..299) {
                val decoded = response.body<AssociationResponse>()
                Pair(decoded.message, decoded.imageBase64)
            } else {
                val errorMessage = response.bodyAsText()
                throw PlateAPIError.ServerError(response.status.value, errorMessage)
            }
        } catch (e: Exception) {
            when (e) {
                is PlateAPIError -> throw e
                else -> throw PlateAPIError.RequestFailed(e)
            }
        }
    }
}
